client/components/jetpack/connection-health/use-check-jetpack-connection-health.ts

```diff
@@ -0,0 +1,46 @@
+import { useQuery } from '@tanstack/react-query';
+import wp from 'calypso/lib/wp';
+import { useDispatch } from 'calypso/state';
+import {
+	setJetpackConnectionHealthy,
+	setJetpackConnectionUnhealthy,
+} from 'calypso/state/jetpack-connection-health/actions';
+
+export const JETPACK_CONNECTION_HEALTH_QUERY_KEY = 'jetpack-connection-health';
+
+export interface JetpackConnectionHealth {
+	is_healthy: boolean;
+	error: string;
+}
+
+interface UseCheckJetpackConnectionHealthOptions {
+	onError?: () => void;
+	onSuccess?: ( data: JetpackConnectionHealth | undefined ) => void;
+}
+
+export const useCheckJetpackConnectionHealth = (
+	siteId: number,
+	options: UseCheckJetpackConnectionHealthOptions
+) => {
+	const dispatch = useDispatch();
+	return useQuery< JetpackConnectionHealth, unknown, JetpackConnectionHealth >( {
+		queryKey: [ JETPACK_CONNECTION_HEALTH_QUERY_KEY, siteId ],
+		queryFn: () =>
+			wp.req.get( {
+				path: `/sites/${ siteId }/jetpack-connection-health`,
+				apiNamespace: 'wpcom/v2',
+			} ),
+		meta: {
+			persist: false,
+		},
+		staleTime: 10 * 1000,
+		onError: options?.onError,
+		onSuccess: ( data ) => {
+			if ( data?.is_healthy ) {
+				dispatch( setJetpackConnectionHealthy( siteId ) );
```

Comment:
Syncing Redux state from RQ is an anti-pattern and should be avoided if possible as it's an easy source for bugs. I came across this piece of code while working on the [RQv5](https://github.com/Automattic/wp-calypso/pull/84338) update. In v5 `on*` handlers are removed from the `useQuery` API (for already mentioned reasons) so we have to migrate all existing usages.

I created #84570 which refactors this code to `useEffect` to unblock the v5 upgrade although it's not ideal as we shouldn't be state-syncing at all. However, looking at the code I can see the Redux selector usages are not few and #81086 introduced complexity on top of that so it's not a straightforward refactor to exclusively use a RQ hook for connection health status.

@wojtekn @kozer @danielbachhuber would any of you be able to look into this and help unmangle usage of this query hook with the Redux state? Ideally we'd either use this query hook wherever we want to retrieve the state of the connection OR migrate to our legacy data layer and not use RQ at all if we want to keep the Redux stuff. I think the former is preferred in general but I'm not fully aware of all the context there may be for choosing Redux state in this case.

Reply:
Thank you!